broker/proposal.go

```diff
@@ -48,70 +65,53 @@ func ProposalFx(w http.ResponseWriter, r *http.Request) (string, interface{}, er
-	guaranteFire := lib.GetDatasetByEnv(origin, models.GuaranteeCollection)
+func proposal(policy *models.Policy) error {
+	log.Println("[proposal] starting bpmn flow...")
 
-	log.Println("[Proposal] starting bpmn flow...")
 	state := runBrokerBpmn(policy, proposalFlowKey)
 	if state == nil || state.Data == nil {
-		log.Println("[Proposal] error bpmn - state not set")
+		log.Println("[proposal] error bpmn - state not set")
 		return errors.New("error on bpmn - no data present")
 	}
-	*policy = *state.Data
+	if state.IsFailed {
+		log.Println("[proposal] error bpmn - state failed")
+		return errors.New("error bpmn - state failed")
+	}
 
-	log.Println("[Proposal] saving proposal to firestore...")
-	policyUid := lib.NewDoc(policyFire)
-	policy.Uid = policyUid
-	err = lib.SetFirestoreErr(policyFire, policyUid, policy)
-	lib.CheckError(err)
+	*policy = *state.Data
 
-	log.Println("[Proposal] saving proposal to bigquery...")
+	log.Printf("[proposal] saving proposal n. %d to bigquery...", policy.ProposalNumber)
 	policy.BigquerySave(origin)
 
-	log.Println("[Proposal] saving guarantees to bigquery...")
-	models.SetGuaranteBigquery(*policy, "proposal", guaranteFire)
-
-	log.Println("[Proposal] end ----------------------------------------------")
-	return err
+	return nil
 }
 
-func setProposalData(policy *models.Policy) {
+func setProposalData(policy *models.Policy) error {
 	log.Println("[setProposalData]")
 
-	policyFire := lib.GetDatasetByEnv(origin, models.PolicyCollection)
-	now := time.Now().UTC()
-
-	policy.CreationDate = now
-	policy.RenewDate = policy.CreationDate.AddDate(1, 0, 0)
-	policy.Status = models.PolicyStatusInitLead
-	policy.StatusHistory = append(policy.StatusHistory, policy.Status)
+	firePolicy := lib.GetDatasetByEnv(origin, models.PolicyCollection)
 
-	numb := GetSequenceProposal("", policyFire)
-	policy.ProposalNumber = numb
-	policy.IsSign = false
-	policy.IsPay = false
-	policy.Updated = now
+	setProposalNumber(policy)
+	policy.Status = models.PolicyStatusProposal
 
-	if policy.ProductVersion == "" {
-		policy.ProductVersion = "v1"
+	if policy.IsReserved {
+		log.Println("[setProposalData] setting NeedsApproval status")
+		policy.Status = models.PolicyStatusNeedsApproval
 	}
 
-	// TODO delete me when PMI is fixed
-	if policy.Name == models.PmiProduct {
-		policy.NameDesc = "Wopta per te Artigiani & Imprese"
-	}
+	log.Printf("[setProposalData] policy status %s", policy.Status)
+
+	policy.StatusHistory = append(policy.StatusHistory, policy.Status)
+	policy.Updated = time.Now().UTC()
 
-	policy.Attachments = &[]models.Attachment{{
-		Name: "Precontrattuale", FileName: "Precontrattuale.pdf",
-		Link: "gs://documents-public-dev/information-sets/" + policy.Name + "/" + policy.ProductVersion + "/Precontrattuale.pdf",
-	}}
+	log.Printf("[setProposalData] saving proposal n. %d to firestore...", policy.ProposalNumber)
+	return lib.SetFirestoreErr(firePolicy, policy.Uid, policy)
```

Comment:
Move the saving procedure to the bpmn handler. This function should set only the data, it is the caller that decides when and where to save it.